src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IListConverter.cs

```diff
@@ -13,7 +13,12 @@ internal sealed class IListConverter<TCollection>
     {
         protected override void Add(in object? value, ref ReadStack state)
         {
-            ((IList)state.Current.ReturnValue!).Add(value);
+            TCollection collection = (TCollection)state.Current.ReturnValue!;
```

Comment:
(FYI ) From the `ArrayList` benchmark results you shared it looks like there is a little improvement due to switching from casting to `IList` to casting to `TCollection` which follows from local microbenchmarks I shared in https://github.com/dotnet/runtime/pull/41223#discussion_r487640117. This comment also applies to other diffs where you switch from `IInterface` to `TCollection`.